compiler/modules/types/list.bal

```diff
@@ -238,11 +265,11 @@ function listInhabited(Context cx, SemType[] members, SemType rest, Conjunction?
         // return !isEmpty(cx, d1) &&  tupleInhabited(cx, [s[0], d1], neg.rest);
         // We can generalize this to tuples of arbitrary length.
         foreach int i in 0 ..< len {
-            SemType ntm = i < negLen ? nt.members[i] : nt.rest;
-            SemType d = diff(members[i], ntm);
+            SemType ntm = i < negLen ? listMemberGet(nt.members, i) : nt.rest;
+            SemType d = diff(listMemberGet(members, i), ntm);
             if !isEmpty(cx, d) {
-                SemType[] s = shallowCopyTypes(members);
-                s[i] = d;
+                ListAtomicTypeMembers s = shallowCopyListAtomicType(members);
+                s.initial[i] = d;
```

Comment:
I would replace these two lines by `ListAtomicMembers s = listAtomicMemberReplace(members, i, d)`: you are creating a new ListAtomicMembers like `members` but with the i-th member replaced by d.